src/content/symbol-store.js

```diff
@@ -14,70 +30,86 @@ export class SymbolStore {
    *                                which will be called whenever we need a symbol table. This method
    *                                needs to return a promise of [addr, syms] (the symbol table).
    */
-  constructor(dbNamePrefix, symbolProvider) {
+  constructor(dbNamePrefix: string, symbolProvider: SymbolProvider) {
     this._symbolProvider = symbolProvider;
-    this._db = new SymbolStoreDB(`${dbNamePrefix}-symbol-tables`);
+    this._db = new SymbolStoreDB(`${dbNamePrefix}-symbol-tables`, 200);
 
     // A set of strings identifying libraries that we have requested
     // symbols for but gotten an error back.
     this._failedRequests = new Set();
 
     // A map with one entry for each library that we have requested (but not yet
     // received) symbols for. The keys are strings (libid), and the values are
-    // promises.
-    this._importingLibs = new Map();
+    // promises that resolve to [ addrs, index, buffer ] symbol tables.
+    this._requestedSymbolTables = new Map();
   }
 
-  _ensureLibraryIsInDB(lib) {
+  _getSymbolTable(lib: Library): Promise<SymbolTableAsTuple> {
     const { debugName, breakpadId } = lib;
     const libid = `${debugName}/${breakpadId}`;
+
     if (this._failedRequests.has(libid)) {
-      return Promise.reject(new Error('We\'ve tried to request a symbol table for this library before and failed, so we\'re not trying again.'));
+      return Promise.reject(
+        new Error('We\'ve tried to request a symbol table for this library before and failed, so we\'re not trying again.'));
     }
-    return this._db.getLibKey(debugName, breakpadId).catch(() => {
-      if (this._importingLibs.has(libid)) {
-        // We've already requested a symbol table for this library and are
-        // waiting for the result, so just return the promise for the existing
-        // request.
-        return this._importingLibs.get(libid);
-      }
 
+    const existingRequest = this._requestedSymbolTables.get(libid);
+    if (existingRequest !== undefined) {
+      // We've already requested a symbol table for this library and are
+      // waiting for the result, so just return the promise for the existing
+      // request.
+      return existingRequest;
+    }
+
+    // Try to get the symbol table from the database
+    const symbolTablePromise = this._db.getSymbolTable(debugName, breakpadId).catch(() => {
       // Request the symbol table from the symbol provider.
-      const symbolTablePromise = this._symbolProvider.requestSymbolTable(lib.debugName, lib.breakpadId);
-      const importPromise = symbolTablePromise.then(symbolTable => {
-        this._importingLibs.delete(libid);
-        return this._db.importLibrary(debugName, breakpadId, symbolTable);
-      }, error => {
-        this._importingLibs.delete(libid);
+      const symbolTablePromise = this._symbolProvider.requestSymbolTable(debugName, breakpadId).catch(error => {
         console.error(`Failed to symbolicate library ${debugName}: ${error}`);
```

Comment:
nit: how about passing the error as second argument ? so that we can have the stack in the console ? (or maybe it's like this on purpose ?)